presto-main/src/main/java/com/facebook/presto/type/DoubleOperators.java

```diff
@@ -237,25 +232,38 @@ public static long hashCode(@SqlType(StandardTypes.DOUBLE) double value)
     @SqlType(StandardTypes.BIGINT)
     public static long saturatedFloorCastToBigint(@SqlType(StandardTypes.DOUBLE) double value)
     {
-        if (value <= MIN_LONG_AS_DOUBLE) {
-            return Long.MIN_VALUE;
-        }
-        if (MAX_LONG_AS_DOUBLE_PLUS_ONE - value <= 1) {
-            return Long.MAX_VALUE;
-        }
-        return DoubleMath.roundToLong(value, FLOOR);
+        return saturatedFloorCastToLong(value, 0x1p63, -0x1p63, Long.MAX_VALUE, Long.MIN_VALUE);
     }
 
     @ScalarOperator(SATURATED_FLOOR_CAST)
     @SqlType(StandardTypes.INTEGER)
     public static long saturatedFloorCastToInteger(@SqlType(StandardTypes.DOUBLE) double value)
     {
-        if (value <= MIN_INT_AS_DOUBLE) {
-            return Integer.MIN_VALUE;
+        return saturatedFloorCastToLong(value, 0x1p31, -0x1p31, Integer.MAX_VALUE, Integer.MIN_VALUE);
+    }
+
+    @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.SMALLINT)
+    public static long saturatedFloorCastToSmallint(@SqlType(StandardTypes.DOUBLE) double value)
+    {
+        return saturatedFloorCastToLong(value, 0x1p15, -0x1p15, Short.MAX_VALUE, Short.MIN_VALUE);
+    }
+
+    @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.TINYINT)
+    public static long saturatedFloorCastToTinyint(@SqlType(StandardTypes.DOUBLE) double value)
+    {
+        return saturatedFloorCastToLong(value, 0x1p7, -0x1p7, Byte.MAX_VALUE, Byte.MIN_VALUE);
+    }
+
+    private static long saturatedFloorCastToLong(double value, double maxValuePlusOneAsDouble, double minValueAsDouble, long maxValue, long minValue)
+    {
+        if (value <= minValueAsDouble) {
```

Comment:
As we discussed because of the precision problem for the MAX_LONG value represented as double.